### PR TITLE
Weights

### PR DESCRIPTION
**Pull request by fgrunewald**
Martini CG mappings are a curious thing; More complex molecules frequently have atoms that are ignored in the mapping (i.e. zero weight) or reweighted. This PR implements a syntax for this. Optionally, nodes can be given weight using`[node;weight]`, where weight may be a float. Hydrogens need not be given as we figure them out later, but canonically speaking, it would be better. Annotation of the weight may occur before or after the chirality operator. A default weight of 1 is given to all nodes. Hydrogen atoms unless explicit get the same weight as the node they are attached to. 

For example, hexadecane but we map only the central two carbons: `{[#C1]|4}.{#C1=[C;0]CC[C;0]}`

### CODE REVIEW

**Comment by pckroon** [comment]
Don't hate it. 
We could also use the 'class' attribute for this (which already exists in smiles). The downsides of that are that it's only integers, and the default is 0; the good news here is that pysmiles doesn't set a class for atoms that don't specify one. 
I'm fine with using `;<float>` instead as well, but I *strongly* suggest they come all the way at the end of an atom specification (and only there).
As for implicit hydrogens, I see 2 options: either a) implicit hydrogens get the same weight as their parent; or b) the weight of the parent is split/distributed over the hydrogens.
 a) gets you more CoG-like behaviour, but CH3 suddenly has more weight (=4) than CH2 (=3), even though you didn't say that.
 b) means that CH3 (=0.25 per atom) and CH2 (=0.33 per atom) both have the same weight, but it's less CoG-like

I think a) is the better option, because CoG is a sane default. Needs to be documented though.

File: cgsmiles/pysmiles_utils.py [hunk 1]
```diff
@@ -83,6 +83,11 @@ def rebuild_h_atoms(mol_graph, keep_bonding=False):
             ref_node = next(mol_graph.neighbors(node))
             mol_graph.nodes[node]["fragid"] = mol_graph.nodes[ref_node]["fragid"]
             mol_graph.nodes[node]["fragname"] = mol_graph.nodes[ref_node]["fragname"]
+        if mol_graph.nodes[node].get("element", "*") == "H":
+            # make sure the weights are copied for implicit h-atoms
+            anchor = list(mol_graph.neighbors(node))[0]
+            weight = mol_graph.nodes[anchor].get("weight", 1)
+            mol_graph.nodes[node]["weight"] = weight
```
Comment (pckroon) [inline cgsmiles/pysmiles_utils.py:90]: How do you know this one was implcit?

**Review by pckroon — CHANGES_REQUESTED** [review]
Code looks fine, just a small question

File: cgsmiles/pysmiles_utils.py (hunk 1 above)
Comment (pckroon) [reply cgsmiles/pysmiles_utils.py:90]: You could say you only allow specified weights for explicit H. You still need to do a little bookkeeping to keep then track of what was explicit or implicit though.

**Comment by fgrunewald** [comment]
@pckroon I prefer to strip them from the smiles string, because in that fashion CGSmiles and SMILES fragments are treated the same. The question is if one would want to allow general annotation for CGSmiles as well. Something like `[#TC5;label]`. I might add that in a seperate PR.

**Comment by pckroon** [comment]
`[#TC5;w=0.5;name=a;attr=42]` you mean? Kind of neat idea.

**Comment by fgrunewald** [comment]
@pckroon yes that's the idea pretty much. In Martini there are so many meta infos, which one might want to add. But again I think a seperate PR would be better

**Comment by fgrunewald** [comment]
ahh yes that is true of course

**Comment by fgrunewald** [comment]
let me sit on this until tomorrow

**Comment by fgrunewald** [comment]
@pckroon here my thoughts:

What I like:

- the syntax makes it easy to store attributes like names; bead types which are often used complementary
- there is no prior assumption to the meaning of a field
- it would allow for a module dealing with these annotations using the same code base / logic

What I don't like:
- another series of equal signs 
- the annotations become more verbose
- it is not clear how attributes propagate to hydrogen atoms; although the sane default is copying the attribute I guess

**Comment by pckroon** [comment]
Agnosticism is a big plus when programming, especially when it comes to syntax design (IMNSHO). It makes it more likely you get it right the first time.

> another series of equal signs
You need some sort of separator between labels and values, so if you want bead annotations this is unavoidable

> the annotations become more verbose
This is indeed a downside. You could argue that, since weights are going to be the most used, they can go without label.
Second thought: treat them like args and kwargs: you can first put positional annotations, then keyword annotations. The application determines the semantics. Use https://docs.python.org/3/library/inspect.html#introspecting-callables-with-the-signature-object

> it is not clear how attributes propagate to hydrogen atoms; although the sane default is copying the attribute I guess
Sounds reasonable. If you want to annotate hydrogens, make them explicit.

**Comment by ricalessandri** [comment]
My 2 cents: This seems neat and useful. About the verbosity concern - since weight=1 will likely be the *implicit* default, verbosity only matters when you need different weight values. And for weights (probably the most common case), `[C;w=0.4]` isn't much longer than `[C;0.4]`. The more explicit syntax actually makes the cgsmiles more human-readable, so I'd lean towards `[C;w=0.4]`. Though I won't oppose the idea of having the weights working (also) without the "w" label: so, if one omits the `w=`, 0.4 is still read as a weight.

**Comment by fgrunewald** [comment]
As decided by popular vote here it is:

We now allow general annotations of the type [keyword=value,keyword=value]. In addition to making the notation more concise, we have positional arguments defined by a dialect. These positional arguments also set default values for example charge: 0 and weight: 1. They also overtake the charge argument. Thus charged beads have to be written as `[#PO4;1.0]` or `[#PO4;c=1.0]`. 

For the CG part, we adopt the CGSmiles base dialect with charges and weights as positional arguments. Therefore, one can write `[#PO4;1.0;0.5]` corresponding to charge1 and weight of 0.5. For a neutral residue `[#SP4;0;0.5]` or `[#SP4;w=0.5]`. 

The notation also allows for example to write:

`[#ION;1.0;type=Na;cg=1;resname=Na]` if one wanted that. 

We also support the annotations for the all-atom part, but we should not advertise it. At the moment one should only specify the weight. Charges should still be specified as `[Na+]` for example. In the long run, when pysmiles gets a minimal parser one can change that and align the annotations. 

So bottom line:

- in OpenSmiles fragments we support weights in addition to the regular SMILES stuff
- in the CG part we support charges and weights as well as any keyword argument

File: cgsmiles/pysmiles_utils.py [hunk 2]
```diff
@@ -83,6 +83,16 @@ def rebuild_h_atoms(mol_graph, keep_bonding=False):
...
+        if mol_graph.nodes[node].get("element", "*") == "H":
+            anchor = list(mol_graph.neighbors(node))[0]
+            # the weight for the hydrogen atom was explicitly set
+            hweights = mol_graph.nodes[anchor].get('hweight', [])
+            if hweights:
+                weight = hweights.pop()
+            # make sure the weights are copied for implicit h-atoms
+            else:
+                weight = mol_graph.nodes[anchor].get("w", 1)
+            mol_graph.nodes[node]["w"] = weight
```
Comment (fgrunewald) [inline cgsmiles/pysmiles_utils.py:95]: @pckroon I think this bit should probably propagate all attributes to hydrogen atoms?!

**Review by pckroon — CHANGES_REQUESTED** [review]

File: cgsmiles/dialects.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    args_found = []
+    kwargs_found = {}
+    if len(string_iteratable) > 0:
+        elements = string_iteratable.split(';')
+        for entry in elements:
+            key_value = entry.split('=')
+            if len(key_value) == 1:
+                args_found.append(key_value[0])
+            else:
+                kwargs_found[key_value[0]] = key_value[1]
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply cgsmiles/dialects.py:35]: @pckroon any suggestions for a better function name?

So you would put `;` as annotation_seperator and `=` as assigment_token or so?

[the remark this answers is not in the document]
Comment (pckroon) [reply cgsmiles/dialects.py:35]: For starters. The bigger "problem" is that I cannot set the atomname to "A=;B". Is it a sensible thing to do? Probably not, but it icks that it's not possible. Similar for attribute names

File: cgsmiles/dialects.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    parameters = []
+    for argname, default_value in default_attributes.items():
+        arg_type = type(default_value)
+        parameters.append(Parameter(argname,
+                                    Parameter.POSITIONAL_OR_KEYWORD,
+                                    default=default_value,
+                                    annotation=arg_type))
+    sig = Signature(parameters)
+    return sig
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply cgsmiles/dialects.py:58]: It's cool in principle, but does it justify the extra complexity? How many people are going to know the tree syntax? A dict is sort of something the average researcher can relate to. I could imagine it as a to-do for later.

File: cgsmiles/dialects.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fgrunewald) [reply cgsmiles/dialects.py:35]: yes but that means one has to faround with regular expressions. I'll give it a go

[the remark this answers is not in the document]
Comment (pckroon) [reply cgsmiles/dialects.py:35]: That's a valid choice. Just making sure it's a choice ;)

File: cgsmiles/dialects.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pckroon) [reply cgsmiles/dialects.py:58]: The user would provide the dialect string, which has the same syntax as a python function definition. The advantage is that you can then have required arguments as well

File: cgsmiles/dialects.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fgrunewald) [reply cgsmiles/dialects.py:35]: Done & Done

File: cgsmiles/dialects.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fgrunewald) [reply cgsmiles/dialects.py:58]: let's table this for later; it relatively far in the deep and not relevant for the immediate syntax.

File: cgsmiles/dialects.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+##########################################################
+#                   KNOWN DIALECTS                       #
+##########################################################
+# this one is for global use
+# it is the base CGSmiles dialect
+GRAPH_BASE = create_dialect({"fragname": "NaN",
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply cgsmiles/dialects.py]: let's go with `CGSMILES_DEFAULT_DIALECT`

File: cgsmiles/dialects.py [hunk 6]
```diff
@@ -86,6 +89,18 @@ def _parse_dialect_string(string_iterable,
...
+    # this should only apply to args know to
+    # the signature
+    remove_keys = []
+    for old_key, new_key in arg_to_fullname.items():
+        if old_key in applied_labels.arguments:
+            applied_labels.arguments[new_key] = applied_labels.arguments[old_key]
+            remove_keys.append(old_key)
+
+    for key in remove_keys:
+        del applied_labels.arguments[key]
```
Comment (pckroon) [inline cgsmiles/dialects.py:102]:
```suggestion
    for old_key, new_key in arg_to_fullname.items():
        if old_key in applied_labels.arguments:
            applied_labels.arguments[new_key] = applied_labels.arguments.pop(old_key)
```

**Review by pckroon — CHANGES_REQUESTED** [review]
Small code cleaning opportunity, looks good otherwise!
